patches/tModLoader/Terraria/ModLoader/ModBossBarStyle.cs

```diff
@@ -0,0 +1,67 @@
+﻿using Microsoft.Xna.Framework.Graphics;
+using Terraria.GameContent.UI.BigProgressBar;
+
+namespace Terraria.ModLoader
+{
+	/// <summary>
+	/// A class that is used to swap out the entire boss bar display system with your own implementation
+	/// </summary>
+	public abstract class ModBossBarStyle : ModType
+	{
+		/// <summary>
+		/// The string that gets saved in the config to represent this ModBossBarStyle. Defaults to "TheMod/TheModBossBarStyle".
+		/// </summary>
+		public string SavedName => $"{Mod?.Name ?? "Terraria"}/{Name}";
```

Comment:
"Defaults to"? Was this supposed to be virtual? If yes - why? If no - you can remove this and change ModType.FullName to have null-coalescing on Mod like here without introducing any problems.

Reply:
copied from ModMenu. I was using ModType.FullName at first and it errored for the non-autoloaded DefaultBossBarStyle where i thought about the same approach you suggested, but didn't act on it

Reply:
Regarding this, Here is the code from ModMenu, including the comment
https://github.com/tModLoader/tModLoader/blob/34fd0def138e1edf2ebf8aa06646dd76955d08d2/patches/tModLoader/Terraria/ModLoader/ModMenu.cs#L19-L22

https://github.com/tModLoader/tModLoader/pull/1034 refers to the ModType PR, that got merged 10 days after the ModMenu one. The obvious problem with using ModType.FullName instead of whatever SavedName is getting currently is that since it's not autoloaded by anything, no mod belongs to it, so it throws.

https://github.com/tModLoader/tModLoader/blob/34fd0def138e1edf2ebf8aa06646dd76955d08d2/patches/tModLoader/Terraria/ModLoader/ModType.cs#L19-L21

As Mirsario pointed out, it could be changed for ModType, would it affect anything? @Rartrin

Reply:
First off, the using "Terraria/Name" is only for vanilla things. We reserve it everywhere else for that and some things would likely break otherwise. Plus, it's always good to be consistent.

Secondly, the real question is why is Mod not getting set. It doesn't matter how it is being loaded. Regardless of if it is added through autoloading or through manual calling, `Load` must be getting called. We can't allow for ModTypes to be exist without a mod and frankly, there is no point in extending from ModType otherwise. If FullName is being called and Mod is null, throw an exception because it means something is likely running in the wrong hook.

Reply:
Currently, there is a `VanillaBossBarStyle` class with `[Autoload(false)]`, and it is initialized in the static constructor of `BossBarLoader`, so it can be added directly to the list of boss bar styles. No registering takes place, so it's not visible to `ModTypeLookup`, nor are basic `ModType` variables properly set.

This design is a straight copy from ModMenu, and I *think* it is done like this to avoid having this style assigned to ModLoaderMod, which gets loaded and unloaded, all while main menu things are happening and possibly checking its value. After all, CB merged it, so he must've had a reason to think this is the correct design.

The current problem obviously is that if some mod would want to check CurrentStyle, and its Mod field, it could possibly be null if the style is vanilla (This same problem exists in the ModMenu implementation aswell).

I could try changing it tomorrow so it belongs to ModLoaderMod, and see if any issues arise.

Reply:
`ModType.Mod` will be null in many cases related to vanilla dummies that never get registered and/or unloaded with mod reloading, and i.e. aren't associated with the internal TML mod. Like with instances of the internal VanillaPlayerDrawLayer and ModMenus. I will commit null coalescing on the main branch in a minute.

Reply:
Then we should solve the issue in an object oriented way rather than hard coding it. `FullName` is working as intended, the problem is that Mod is null.

Reply:
Due to tml commit https://github.com/tModLoader/tModLoader/commit/f79079253b05c0d734839fb212eb234fd480e508, can this now be resolved? @Rartrin